terraform/03_gke_cluster.tf

```diff
@@ -117,33 +128,62 @@ resource "null_resource" "current_project" {
   }
 }
 
-#resource "null_resource" "sleeping_subprocess" {
-#  provisioner "local-exec" {
-#      command = "sleep 60 >./stdout.log 2>./stderr.log & echo \"sleeping in PID\" $!"
-#  }
-#
-#  depends_on = [google_container_cluster.gke]
-#}
-
-# Setting kubectl context to currently deployed GKE cluster
-resource "null_resource" "set_gke_context" {
-  provisioner "local-exec" {
-    command = "gcloud container clusters get-credentials cloud-ops-sandbox --zone ${element(random_shuffle.zone.result, 0)} --project ${data.google_project.project.project_id}"
-  }
+# Create GSA to allow K8S services to access Google APIs
+resource "google_service_account" "set_gsa" {
+  account_id   = "gke-sa"
+  display_name = "gsa"
+  project = data.google_project.project.project_id
 
   depends_on = [
-    google_container_cluster.gke, 
+    google_container_cluster.gke,
     null_resource.current_project
   ]
 }
 
+# Create GSA/KSA binding: let IAM auth KSAs as a svc.id.goog member name
+resource "google_service_account_iam_binding" "set_gsa_binding" {
+  service_account_id = google_service_account.set_gsa.name
+  role = "roles/iam.workloadIdentityUser"
+
+  members = [
+    "serviceAccount:${data.google_project.project.project_id}.svc.id.goog[default/default]"
+  ]
+
+  depends_on = [google_service_account.set_gsa]
+}
+
+# Annotate KSA
+resource "null_resource" "annotate_ksa" {
+  triggers = {
+    cluster_ep = google_container_cluster.gke.endpoint  #kubernetes cluster endpoint
+  }
+
+  provisioner "local-exec" {
+    command = <<EOT
+      gcloud container clusters get-credentials cloud-ops-sandbox --zone ${element(random_shuffle.zone.result, 0)} --project ${data.google_project.project.project_id}
+      kubectl annotate serviceaccount --namespace default default iam.gke.io/gcp-service-account=${google_service_account.set_gsa.email}
+    EOT
+  }
+
+  depends_on = [google_service_account_iam_binding.set_gsa_binding]
+}
+
+# Set editor role for GSA to enable Cloud Trace functionality in cluster
+resource "null_resource" "set_editor" {
```

Comment:
too much power.
Can we use roles/cloudtrace.agent?
If not, then maybe roles/cloudtrace.user?

Reply:
Sure. I'll extend the permissioning to cover a monitoring role as well. Otherwise I get: 
`com.google.api.gax.rpc.PermissionDeniedException: io.grpc.StatusRuntimeException: PERMISSION_DENIED: Permission monitoring.timeSeries.create denied (or the resource may not exist).`

Reply:
It might need to be `cloudtrace.admin` . the other two roles still yielded access errors

Reply:
Actually, we need to keep it as `editor` role. My inline comment was wrong/misleading. Trace isn’t the only set of permissions this GSA needs, esp during deployment. Given the WorkloadId setup, this GSA now replaces Compute Engine default service account, which had editor role. So we can keep parity.

Btw, I did try and demote GSA roles to just trace.admin & monitoring.editor. During deploy, all other Google APIs became disabled (iam/error/all permissions were missing).

Reply:
OK, sounds good, let's keep it editor role. Thanks for checking on this.
Worth updating the comment in line 171 to reflect that it's not only trace that requires editor.